pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -895,24 +897,27 @@ private void pendingBatchReceiveTask(Timeout timeout) throws Exception {
     protected void triggerListener() {
         // Trigger the notification on the message listener in a separate thread to avoid blocking the networking
         // thread while the message processing happens
-        Message<T> msg;
-        do {
-            try {
-                msg = internalReceive(0, TimeUnit.MILLISECONDS);
+        try {
+            // Control executor to call MessageListener one by one.
+            if (executorQueueSize.get() < 1) {
+                Message<T> msg = internalReceive(0, TimeUnit.MILLISECONDS);
```

Comment:
we can mark `msg` as `final` and delete the `finalMsg` variable.

Reply:
Nice review, thanks.